DirectProgramming/DPC++FPGA/include/streaming_cholesky.hpp

```diff
@@ -0,0 +1,254 @@
+                  "The pipe must be able to contain at least one element");
+
+    // Set the computation type to T or ac_complex<T> depending on the value
+    // of is_complex
+    using TT = std::conditional_t<is_complex, ac_complex<T>, T>;
+
+    constexpr int kColumns = rows;
+
+    // Number of lower-left elements in the L output matrix
+    constexpr int kLMatrixSize = kColumns * (kColumns + 1) / 2;
+
+    // Compute Cholesky decompositions as long as matrices are given as inputs
+    while (1) {
+      // Break memories up to store pipe_size elements per bank
+      constexpr short kBankwidth = pipe_size * sizeof(TT);
+      constexpr unsigned short kNumBanks = rows / pipe_size;
+
+      // When specifying numbanks for a memory, it must be a power of 2.
+      // Unused banks will be automatically optimized away.
+      constexpr short kNumBanksNextPow2 =
+          fpga_tools::Pow2(fpga_tools::CeilLog2(kNumBanks));
+
+      [[intel::numbanks(kNumBanksNextPow2)]]  // NO-FORMAT: Attribute
+      [[intel::bankwidth(kBankwidth)]]        // NO-FORMAT: Attribute
+      [[intel::private_copies(4)]]            // NO-FORMAT: Attribute
+      [[intel::max_replicates(1)]]            // NO-FORMAT: Attribute
+      TT a_load[rows][kColumns];
+
+      // Two copies of L to be able to load two complete rows per iteration
+      // Multiple private copies to be able to overlap multiple loop
+      // iterations
+      [[intel::private_copies(4)]]  // NO-FORMAT: Attribute
+      TT l_result_compute[rows][kColumns];
+      [[intel::private_copies(4)]]  // NO-FORMAT: Attribute
+      TT l_result_compute_copy[rows][kColumns];
+
+      [[intel::private_copies(2)]]  // NO-FORMAT: Attribute
+      TT l_result[kLMatrixSize];
+
+      // Copy a matrix from the pipe to a local memory
+      // Number of pipe reads of pipe_size required to read a full column
+      constexpr int kExtraIteration = ((rows % pipe_size) != 0) ? 1 : 0;
+      constexpr int kLoopIterPerColumn = (rows / pipe_size) + kExtraIteration;
+      // Number of pipe reads of pipe_size to read all the matrices
+      constexpr int kLoopIter = kLoopIterPerColumn * kColumns;
+      // Size in bits of the loop iterator over kLoopIter iterations
+      constexpr int kLoopIterBitSize =
+          fpga_tools::BitsForMaxValue<kLoopIter + 1>();
+
+      [[intel::initiation_interval(1)]]  // NO-FORMAT: Attribute
+      for (ac_int<kLoopIterBitSize, false> li = 0; li < kLoopIter; li++) {
+        fpga_tools::NTuple<TT, pipe_size> pipe_read = AIn::read();
+
+        int write_idx = li % kLoopIterPerColumn;
+
+        fpga_tools::UnrolledLoop<kLoopIterPerColumn>([&](auto k) {
+          fpga_tools::UnrolledLoop<pipe_size>([&](auto t) {
+            if (write_idx == k) {
+              if constexpr (k * pipe_size + t < kColumns) {
+                a_load[li / kLoopIterPerColumn][k * pipe_size + t] =
+                    pipe_read.template get<t>();
+              }
+            }
```

Comment:
Not sure it really matters, but maybe the if and if constexpr should we swapped? So we generate no code if the if-constexpr is false? The optimizer will likely get rid of the empty though.